File: elixir/lib/advent_of_code/day_06.ex
defmodule AdventOfCode.Day06 do
  def count(set) do
    Enum.reduce(set, 0, &(&2 + MapSet.size(&1)))
  end

  def part1(input) do
    input
    |> String.split("\n\n")
    |> Enum.map(&to_charlist(&1))
    |> Enum.map(&MapSet.new(&1))
    |> Enum.map(&MapSet.delete(&1, 10))
    |> count
  end

  def part2(input) do
    input
    |> String.split("\n\n")
    |> Enum.map(&String.split(&1))
    |> Enum.map(fn x -> Enum.map(x, &to_char_list(&1)) end)
    |> Enum.map(fn x -> Enum.map(x, &MapSet.new(&1)) end)
    |> Enum.map(fn x -> Enum.reduce(x, &MapSet.intersection(&1, &2)) end)
    |> count
  end
end

File: haskell/src/Day06.hs
module Day06 where

import Data.List.Split
import qualified Data.Set as Set

type GroupAnswer = Set.Set Char

count :: [GroupAnswer] -> Int
count = foldl (\acc x -> acc + Set.size x) 0

part1 :: String -> Int
part1 = count . inputToAnyone

part2 :: String -> Int
part2 = count . inputToEveryone

inputToEveryone :: String -> [GroupAnswer]
inputToEveryone txt =
  let group = splitOn "\n\n" txt
      groupAnswer = map (map Set.fromList . words) group
      everyGroup = map (foldl1 Set.intersection) groupAnswer
   in everyGroup

inputToAnyone :: String -> [GroupAnswer]
inputToAnyone txt =
  let groups = splitOn "\n\n" txt
      groupAnswer = map (Set.delete '\n' . Set.fromList) groups
   in groupAnswer
